src/cloudWatchLogs/registry/logStreamRegistry.ts

```diff
@@ -190,23 +195,37 @@ export async function filterLogEventsFromUriComponents(
     logGroupInfo: CloudWatchLogsGroupInfo,
     parameters: CloudWatchLogsParameters,
     nextToken?: string
-): Promise<CloudWatchLogsResponse> {
+): Promise<CloudWatchLogsResponse | void> {
     const client = new DefaultCloudWatchLogsClient(logGroupInfo.regionName)
 
-    const response = await client.filterLogEvents({
+    const timeout = new Timeout(10000)
+    showMessageWithCancel(`Loading log data from group ${logGroupInfo.groupName}`, timeout)
+    const responsePromise = client.filterLogEvents({
         logGroupName: logGroupInfo.groupName,
         filterPattern: parameters.filterPattern,
         nextToken,
         limit: parameters.limit,
     })
-
-    // Use heuristic of last token as backward token and next token as forward to generalize token form.
-    // Note that this may become inconsistent if the contents of the calls are changing as they are being made.
-    // However, this fail wouldn't really impact customers.
-    return {
-        events: response.events ? response.events : [],
-        nextForwardToken: response.nextToken,
-        nextBackwardToken: nextToken,
+    let response
+    try {
+        response = await waitTimeout(responsePromise, timeout)
```

Comment:
This is not exactly the "right" approach but considering we kept it simple and isolated it's fine.
Cancellation is a pretty advanced topic and it's easy to make your logic a lot buggier/messier than it should be.

We should do this as well:
```ts
const response = await waitTimeout(responsePromise, timeout, { allowUndefined: false })
```

Reply:
What would be the better way? Is it a completely different approach or some small changes? What are the disadvantages of this approach?